initial port to wasm

ed2k opened the pull request:
bazel build, tested under macos
$ bazel build --config=wasm //examples:solve_board_wasm

test
$ node bazel-bin/examples/solve_board.js

```
SolveBoard, hand 1: solutions 3 OK, solutions 2 OK
--------------------------------------------------
            QJ6
            K652
            J85
            T98
AT942                   873
AQ4                     J97
32                      AT764
KJ3                     Q4
            K5
            T83
            KQ9
            A7652


solutions == 3

  card suit   rank   equals score 
     0 D      5             5     
     1 D      8             5     
     2 D      J             5     
     3 C      T      98     5     
     4 S      6             5     
     5 S      Q      J      5     
     6 H      2             4     
     7 H      6      5      4     
     8 H      K             4     

solutions == 2

  card suit   rank   equals score 
     0 D      5             5     
     1 D      8             5     
     2 D      J             5     
     3 C      T      98     5     
     4 S      6             5     
     5 S      Q      J      5     

SolveBoard, hand 2: solutions 3 OK, solutions 2 OK
--------------------------------------------------
            AK96
            KQ8
            A98
            K63
87                      QJT5432
A62                     T
QJT4                    6
AT75                    QJ82
            -
            J97543
            K7532
            94


solutions == 3

  card suit   rank   equals score 
     0 C      2             4     
     1 C      8             4     
     2 C      Q      J      4     
     3 H      T             3     
     4 D      6             3     
     5 S      Q      JT     3     
     6 S      5      432    2     

solutions == 2

  card suit   rank   equals score 
     0 C      2             4     
     1 C      8             4     
     2 C      Q      J      4     

SolveBoard, hand 3: solutions 3 OK, solutions 2 OK
--------------------------------------------------
            73
            QJT
            AQ54
            T752
AKJ9842                 QT6
K                       876
T8                      KJ9
J93                     AQ84
            5
            A95432
            7632
            K6


solutions == 3

  card suit   rank   equals score 
     0 H      A             3     
     1 D      3      2      3     
     2 D      7      6      3     
     3 S      5             3     
     4 H      5      432    2     
     5 H      9             2     
     6 C      6             1     
     7 C      K             1     

solutions == 2

  card suit   rank   equals score 
     0 H      A             3     
     1 D      3      2      3     
     2 D      7      6      3     
     3 S      5             3     
```

ed2k committed:
initial port to wasm

src> make && make install
examples> make && node AnalysePlayBin.js

ed2k committed:
remove .js and .wasm output files

ed2k committed:
fix typo

ed2k committed:
clean up unnecessary change

ed2k committed:
add dds wasm example

S-N-O-R-L-A-X commented:
Hi, I am interested in your contribution. But I am using windows and not familiar with these things. Could you share a compiled wasm file? Many thanks.

ed2k committed:
Merge remote-tracking branch 'origin/develop' into wasm

ed2k committed:
initial wasm build migrate to bazel

ed2k committed:
auto find em++, analyze still have hand 2 ERROR

zzcgumn commented on src/Makefiles/Makefile_wasm:
```diff
@@ -0,0 +1,115 @@
+
+# You can configure the following:
+
+# 1. The threading systems that you want in the library.
```
This is not accurate anymore. Standard library and jthread is the only available option.

zzcgumn commented on src/Makefiles/Makefile_wasm:
```diff
@@ -0,0 +1,115 @@
+# If you need to add something for a threading system, this is
+# the place.
+
+CC_BOOST	= /usr/local/Cellar/boost/1.56.0
```
To my knowledge there is no need for BOOST anymore.

zzcgumn commented on examples/solve_board.cpp:
```diff
@@ -33,7 +33,7 @@ auto main() -> int
 
-#if defined(__linux) || defined(__APPLE__)
+#if defined(__linux) || defined(__APPLE__)  || defined(__WASM__)
   SetMaxThreads(0);
```
This should be a no-op by now.

ed2k replied:
will try later

ed2k committed:
Merge remote-tracking branch 'origin/develop' into wasm

ed2k committed:
Remove Makefile_wasm

old Make system gone

ed2k committed:
remove obsolete make file

ed2k committed:
Add emsdk WASM toolchain and wasm rules

Enable hermetic Emscripten via the emsdk toolchain and switch project WASM config to use it. .bazelrc now configures the emsdk platform/cpu for --config=wasm; BUILD.bazel's wasm config_setting uses values.cpu=wasm. MODULE.bazel adds the emsdk dependency and registers its toolchains. CPPVARIABLES.bzl updates WASM cppopts (adds -fexceptions, removes -mtune=generic). examples/BUILD.bazel now loads emsdk wasm_rules and replaces custom em++ genrules with wasm_cc_binary targets, and updates WASM link options (exceptions and memory flags).

ed2k committed:
Docs: use Bazel hermetic Emscripten for WASM

Revise wasm_build.md to remove manual Emscripten install instructions and note that Bazel will automatically download and configure a hermetic emscripten toolchain. Update BUILD.bazel example to use values={"cpu":"wasm"} and mark the code block as python. Adjust CPPVARIABLES and flags (remove -mtune=generic, add -fexceptions). Update .bazelrc examples to use @emsdk//:platform_wasm, add host_cxxopt and compilation_mode=opt. Remove the em++ not found troubleshooting and the recommendation to make em++ explicitly available in CI. Update the compilation flags table to include -fexceptions, -sALLOW_MEMORY_GROWTH=1 and -sINITIAL_MEMORY=268435456.

ed2k commented on examples/dds.cpp:
```diff
@@ -0,0 +1,148 @@
+#include <string.h>
+#include "../include/dll.h"
+#include "../src/PBN.h"
+#include "hands.h"
```
removed

ed2k committed:
WASM: rename AnalysePlayBin, update docs & cleanup

Rename analyse_play_bin target to AnalysePlayBin and update WASM binary target names and filegroup references; adjust analyse_play_bin.cpp error handling for WASM builds to avoid ErrorMessage and print error code instead; remove deprecated examples/dds.cpp. Also remove -Wno-character-conversion from DDS CPPOPTS. Revise docs/wasm_build.md to use the new WASM target names, simplify available targets list, and trim/clarify WASM-related notes and output instructions.

ed2k commented on examples/analyse_play_bin.cpp:
```diff
@@ -62,7 +62,9 @@ auto main() -> int
 
     if (res != RETURN_NO_FAULT)
     {
+#ifndef __WASM__
       ErrorMessage(res, line);
+#endif
       printf("DDS error: %s\n", line);
     }
```
done
